client/my-sites/controller.js

```diff
@@ -210,45 +226,54 @@ module.exports = {
 			return next();
 		}
 
-		// If there's a valid site from the url path
-		// set site visibility to just that site on the picker
-		if ( sites.select( siteID ) ) {
-			const selectionComplete = onSelectedSiteAvailable( context );
-
-			// if there was a redirect, we should terminate processing of next routes
-			// and let the redirect proceed
-			if ( ! selectionComplete ) {
-				return;
-			}
-		} else {
-			// if sites has fresh data and siteID is invalid
-			// redirect to allSitesPath
-			if ( sites.fetched || ! sites.fetching ) {
-				return page.redirect( allSitesPath );
+		// ensure we have fetched sites
+		const sitesFetchedPromise = new Promise( ( resolve, reject ) => {
+			if ( ! sites.fetched && sites.fetching ) {
+				sites.once( 'change', () => sites.fetched ? resolve() : reject() );
+			} else if ( currentUser.visible_site_count !== sites.getVisible().length ) {
+				sites.initialized = false;
+
+				const waitingNotice = notices.info( i18n.translate( 'Finishing set up…' ), { showDismiss: false } );
+				sites.once( 'change', () => {
+					notices.removeNotice( waitingNotice );
+					sites.fetched ? resolve() : reject();
+				} );
+
+				sites.fetch();
+			} else {
+				return resolve();
 			}
-
-			let waitingNotice;
-			const selectOnSitesChange = () => {
-				// if sites have loaded, but siteID is invalid, redirect to allSitesPath
-				if ( sites.select( siteID ) ) {
-					sites.initialized = true;
```

Comment:
`sites.initialized` is no longer set to true if the site can be selected in this diff. Is this intentional? I'm hesitant to change the behavior of this because it could have far-reaching implications.